File: Person_to_exit_projekt/grid.py
"""
Author: Sofie Konvalinová
Date: 18-11-2025
Email: konvasof@cvut.cz
Description:
    Třída pro správu statické topologie mřížky. 
    Zpracovává vstupní mapy ve formátu (header + ASCII mřížka) 
    a ukládá pevné zdi do optimalizované množiny. Poskytuje robustní 
    metody pro kontrolu průchodnosti a hranic, které využívá PathFinder
 """

class GridMap:

    def __init__(self, raw_map_content: str):
        self.width = 0
        self.height = 0
        self.walls = set()  # Uložíme souřadnice (x, y), kde je zeď
        self.exit_point = None # Cíl nastavíme později     
        self._parse_map(raw_map_content)

    def _parse_map(self, content: str):
        lines = content.strip().split('\n')      
        # Čekáme na začátek mřížky 
        parsing_grid = False
        y = 0 # Řádek v mřížce
        
        for line in lines:
            line = line.strip() # ostrání bílé znaky
            if not line:
                continue
            
            # Parsování hlavičky
            if line.startswith('height'):
                self.height = int(line.split()[1])
                continue
            elif line.startswith('width'):
                self.width = int(line.split()[1])
                continue
            elif line.startswith('map'):
                parsing_grid = True # Začínáme parsovat mřížku
                continue
            
            # Parsování samotné mřížky
            if parsing_grid:
                # Projdeme znaky v řádku
                for x, char in enumerate(line):
                    if char == '@': # @ i T se často berou jako zdi
                        self.walls.add((x, y))
                y += 1

    def set_exit(self, x: int, y: int):
        # Nastaví cíl
        # Nesmí být ve zdi ani mimo mapu
        if not self.in_bounds(x, y):
            raise ValueError(f"Exit {x},{y} je mimo mapu!")
        if (x, y) in self.walls:
            raise ValueError(f"Exit {x},{y} je ve zdi!")
        self.exit_point = (x, y)

    def in_bounds(self, x: int, y: int) -> bool:
        # Zkontroluje, zda jsou souřadnice uvnitř mapy
        return 0 <= x < self.width and 0 <= y < self.height

    def is_walkable(self, x: int, y: int) -> bool:
        # Zkontroluje, zda je pozice průchozí (uvnitř mapy a není zeď)
        return self.in_bounds(x, y) and (x, y) not in self.walls
    

File: Person_to_exit_projekt/my_solver/src/Grid.cpp
#include "Grid.h"
#include <sstream>
#include <iostream>

GridMap::GridMap(const std::string& raw_map_content) {
    _parse_map(raw_map_content);
}

void GridMap::_parse_map(const std::string& content) {
    std::stringstream ss(content);
    std::string line;
    bool parsing_grid = false;
    int y = 0;

    while (std::getline(ss, line)) {
        if (line.empty()) continue;
        
        if (!line.empty() && line.back() == '\r') line.pop_back();

        if (line.find("height") == 0) {
            height = std::stoi(line.substr(7));
        } else if (line.find("width") == 0) {
            width = std::stoi(line.substr(6));
        } else if (line.find("map") == 0) {
            parsing_grid = true;
            continue;
        } else if (parsing_grid) {
            for (int x = 0; x < line.length(); ++x) {
                if (line[x] == '@' || line[x] == 'T') { 
                    walls.insert({x, y});
                }
            }
            y++;
        }
    }
}

void GridMap::set_exit(int x, int y) {
    if (!in_bounds(x, y)) {
        std::cerr << "Exit " << x << "," << y << " is outside the map." << std::endl;
        return;
    }
    // Ujistíme se, že exit není ve zdi
    walls.erase({x, y});
    exit_point = {x, y};
}

bool GridMap::in_bounds(int x, int y) const {
    return x >= 0 && x < width && y >= 0 && y < height;
}

bool GridMap::is_walkable(int x, int y) const {
    return in_bounds(x, y) && walls.find({x, y}) == walls.end();
}
